Rate limit devices per user on short window

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
A 406 http response is returned when this limit is reached, which allows us to send a custom, non-translatable message. However given the mobile worker will be connected to the internet, I'm optimistic they will be able to translate it or send it to someone who can make better use of the error message if needed. While a 429 response would have fit here as well and is translatable, the message is too vague in my opinion. This way we can ensure we are communicating to the user that it is usage for _this specific user_ that is too high.

![image](https://github.com/user-attachments/assets/2daa85f7-2f66-4b9c-9dd4-d95b1f8dcf23)

Based on [android code](https://github.com/dimagi/commcare-android/blob/6604f7021697ebc4a1e98a85e5c9bb32d6104719/app/src/org/commcare/utils/FormUploadUtil.java#L217-L234), I'm pretty sure this will be displayed similarly on mobile, but will confirm with a physical device. **Update**: Confirmed this looks good on mobile device.
## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-16355

Initial PR: https://github.com/dimagi/commcare-hq/pull/35515

**Note for reviewers**
By default, the device rate limiter is not enabled, but the code is in place to have the ability to enable this via `update-config`. Review this code as if it were being enabled immediately, but note that there will be time after this is deployed to collect metrics in datadog to determine how the current limits fit into current usage. 

**Summary**

The changes in this PR track usage for each user action that leads to updating user metadata which are:
- restores
- form submissions
- heartbeats

If the combined usage on these endpoints reaches 10 unique device IDs within a fixed minute time window, any other requests for that user within that minute from a new device ID are rejected. The exceptions are if a device ID is None or blank, or if the device ID is from Web Apps.

**Understanding past usage**

The easiest way to get a rough sense of current usage at a larger scale is to look at the `SyncLogSQL` table and see how many different devices are used by a single user in different windows of time. This fails to capture _all_ activity like form submissions, but gives a good sense of restore activity for a specific user. I've periodically looked at this over the last month for different days/weeks for day windows and have seen that these numbers are relatively consistent, with the exception of December 6th.

The smaller the time window, the less useful this data is, but gives a sense of the difference. The datadog metrics included in this PR will give a much better sense of usage at the minute window level.

***1 day window***
December 6th: [2781, 349, 346, 328, 54, 46, ...]  # the top 4 device counts are from the problematic domain
January 15th: [420, 224, 42, 41, 40, ...]

***Random 1 hour window***
December 6th: [776, 52, 36, 24, 5, ...]
January 17th: [17, 7, 5, 4, 3, ...]

***Random 1 minute window***
December 6th: [34, 3, 1, ...]
January 17th: [2, 1, ...]

Again, I don't think it is worth diving too deep into these numbers as the metric will be far more useful and accurate.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Based on [Clayton's comment](https://dimagi.atlassian.net/browse/SAAS-16355?focusedCommentId=360266), here are the considerations outlined for this change.

#### How will users become aware of this change?
Assuming that this limit is currently being reached or exceeded (we won't know until metrics are collected), here is how users will become aware.

The mobile users will attempt an action like a sync or form submission, and receive an error message explaining, in english, that current usage for this user is too high and that they should try again in a minute. Based on numbers pulled for devices used per user in a day (hardly ever exceeds 100 devices), in most cases where this limit is hit, trying again in a minute should succeed. If however there are thousands of devices attempting to take an action at roughly the same time, it will be a painful process to have every device be able to restore or submit successfully. We suspect this scenario is limited to large events like trainings. 

#### How are other legitimate use cases impacted?
Other use cases outside of simultaneous usage of a user across different physical devices that can lead to multiple device ids for one user are:
1. **Uninstalling and reinstalling the CommCare app**
  It seems nearly impossible to uninstall and reinstall CommCare 9 times in one minute, let alone logging into your user and triggering a restore or submission, so this scenario should not be a concern.

2. **Clearing user data**
   Clearing user data is easier to do frequently, and could lead to frequent device ID changes, but again, a user would have to do this 9 times in one minute. That seems well past the point of debugging an issue.

3. **Multiple CommCare app installations**
  Running with multiple installations of CommCare seems to be the most plausible for actually sending off 10 requests from 10 different installations for one user, but just because it is plausible doesn't mean it makes sense. I'm not too familiar with the use cases for different installations, but I imagine it is testing behavior between different mobile app versions or something along those lines, which should be limited to a few different device IDs at most.

#### Does this impact our offering?
I would say this does not impact our offering, but does make it more painful to use the product in a way we did not intend for (a significant number of devices for one user). This is a temporary rate limit for what we, SaaS, consider to be a reasonable number of devices per user in a minute window.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Created tests for the DeviceRateLimiter class (`corehq.apps.users.tests.test_device_rate_limiter.TestDeviceRateLimiter`)

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Create limiter to track device usage for each user

gherceg committed:
Return 406 response when device is rate limited

gherceg committed:
Added datadog metrics for rate limiter

gherceg committed:
Add setting to enable/disable device rate limiter

gherceg committed:
Fix failing test using mock

TestAuditLoggingForFormSubmission::test_api_user_api_endpoint failed due
to attempting to pass a Mock object to convert_xform_to_json

gherceg commented:
Moving the remaining TODOs to the JIRA ticket:
> - [ ] Create list of projects/users at risk of hitting this limit
> - [ ] Create and link confluence page detailing this limit

The first one will be easier to do based on the datadog metric, and I can work on the confluence page while this is in review.

snopoke reviewed (approved):
LGTM

millerdev commented on corehq/apps/users/tests/test_device_rate_limiter.py:
```diff
@@ -0,0 +1,62 @@
+            device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id')
+            self.assertTrue(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'new-device-id'))
+            frozen_time.move_to("2024-12-10 12:06:15")
+            self.assertFalse(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'new-device-id'))
```
I'm not understanding how this works. Does `frozen_time.move_to(...)` affect the redis concept of time? Also, the time window was only moved 32 seconds past 12:05:43, so it seems like it should not have reset yet even if it did affect the redis TTL.

gherceg replied:
Perhaps my comment in https://github.com/dimagi/commcare-hq/pull/35613#discussion_r1924437377 sheds light on this as well, but `_get_redis_key` bakes the timestamp down to the current minute into it, so we don't need redis' concept of time to change, we just need to move to a new fixed window (i.e. the next minute). This is why I chose seemingly random times between these two minutes that were less than a minute apart as well.

millerdev commented on corehq/apps/users/device_rate_limiter.py:
```diff
@@ -0,0 +1,93 @@
+
+        key = self._get_redis_key(user_id)
+
+        if not self._exists(key):
+            self._track_usage(key, device_id, is_key_new=True)
+            return False
+
+        if self._device_has_been_used(key, device_id):
+            return False
+
+        device_count = self._device_count(key)
+        if device_count < self.device_limit_per_user(domain):
+            self._track_usage(key, device_id)
```
I think this will hit redis at least 3 times and at most 5 times.

The number of hits could be reduced to 1 or 2, and the operation made atomic with a pipeline

```py
pipe = client.pipeline()
pipe.exists(key)
pipe.sadd(key, device_id)
pipe.scard(key)
pre_existing, added, count = pipe.execute()
if pre_existing:
    return added and count >= devices_allowed
client.expire(key, DEVICE_SET_CACHE_TIMEOUT)
return False
```

gherceg replied:
Thanks for mentioning pipelines. Didn't know about this. As we discussed, this code as is would change behavior from what is desired currently. I adopted pipelines in c90a5bd33b991b78260d361c8ec1d1f94c23fbb2 to help minimize connections to redis and ended up with at most 2 calls to redis while keeping the current behavior.

gherceg commented on corehq/apps/users/device_rate_limiter.py:
```diff
@@ -0,0 +1,79 @@
+
+    def _device_has_been_used(self, redis_key, device_id):
+        # check if device_id is member of the set for this key
+        return self.client.srem(redis_key, device_id)
```
Oh wow yeah this is definitely unintentional. `sismember` is what I want. Thanks for highlighting.

gherceg replied:
Resolved in 384f700fb596d8b1f0f363bf54e56e69039af8d9

gherceg commented on corehq/apps/users/device_rate_limiter.py:
```diff
@@ -0,0 +1,79 @@
+    def _track_usage(self, redis_key, device_id, is_key_new=False):
+        self.client.sadd(redis_key, device_id)
+        if is_key_new:
+            self.client.expire(redis_key, DEVICE_SET_CACHE_TIMEOUT)
```
A possibly important distinction is that this isn't a sliding window, but a fixed window. So each redis key has the current timestamp down to the minute baked into it, meaning a key is only relevant for _at most_ one minute. The two minute timeout ensures there isn't any funny business at the boundary between the current minute and the next minute (it could probably just be 70 seconds or something like that).

> I think this means there is a race condition where a device that successfully sync'd in the previous period may be locked out in the next because other devices that were previously locked out used up the quota in the current period

I somewhat follow, but am not sure if you still have a question here after my answer above. With a fixed window, devices locked out in the previous minute wouldn't have an impact on the quota for the next minute, but generally speaking, just because a device was allowed one minute does not guarantee that it will be allowed in the next (which is intentional).

millerdev replied:
This makes sense. I did not discover the 1-minute-baked-into-key feature reading your explanation. What do you think of adding a note about that to a docstring?

gherceg committed:
Use sismember, not srem redis function

Reproduced issue this caused with test:

device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id')
device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id')
device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'new-device-id')
self.assertFalse(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id'))

which failed since the second rate_limit_device call removed
'existing-device-id', and the next call with 'new-device-id' effectively
takes its place in the allowed device list.

gherceg committed:
Use pipelines to reduce calls to redis

millerdev commented on corehq/apps/receiverwrapper/views.py:
```diff
@@ -137,6 +138,17 @@ def _process_form(request, domain, app_id, user_id, authenticated,
+        # let normal response handle invalid xml
+        pass
+    else:
+        device_id = form_json.get('meta', {}).get('deviceID')
```
@gherceg pointed out in an offline discussion that `instance` is a byte string here, not a form object as it is in `SubmissionPost` later on.

It would be nice to pass `form_json` on from here to anything else that subsequently needs parsed form JSON to avoid having to re-parse in those places.

gherceg replied:
Attempted in d3e70de19e97fed8c6067702111a061bc9068245

millerdev commented on corehq/apps/users/device_rate_limiter.py:
```diff
@@ -0,0 +1,93 @@
+from corehq.util.metrics import metrics_counter, metrics_histogram
+
+DEVICE_RATE_LIMIT_MESSAGE = "Current usage for this user is too high. Please try again in a minute."
+DEVICE_SET_CACHE_TIMEOUT = 2 * 60  # 2 minutes
```
This constant was the main thing that drew my attention when I was looking for the rate limit timeout, and it didn't occur to me that the key generation function was the more important place to look for that. What do you think of adding a comment here noting that this is purposefully longer than the 1-minute fixed window enforced by `DeviceRateLimiter._get_redis_key`?

gherceg replied:
4c8c734ca860e30dca73c03cee2e9625f9748694

gherceg committed:
Modify bucket boundaries

gherceg committed:
Pass json converted instance into SubmissionPost

This avoids unnecessarily converting the instance xml to json multiple
times

gherceg commented:
I'm not proud of d3e70de19e97fed8c6067702111a061bc9068245, but wanted to make as minimal of a chance as I could to reduce risk.

gherceg committed:
Add comment explaining rationale for 2 minute timeout

gherceg committed:
Fix bug tracking wrong user id

gherceg committed:
Only use login as user if available

gherceg committed:
Fix metric to report device count

A histogram was the wrong choice for this metric since we report each
individual usage as it happens, not the total number at the end of a
window.

gherceg committed:
Fix bug introduced with previous change

Don't code while you're sick

millerdev commented on corehq/form_processor/parsers/form.py:
```diff
@@ -96,13 +98,14 @@ def _create_new_xform(domain, instance_xml, attachments=None, auth_context=None)
     assert attachments is not None
-    form_data = convert_xform_to_json(instance_xml)
-    if not form_data.get('@xmlns'):
+    if not instance_json:
```
nit: Not sure if `instance_json` could ever be an empty dict, but it seems best to be strict and only convert XML if it is null.
```suggestion
    if instance_json is not None:
```

gherceg replied:
963cb4275a469669a08dd7ce1056e3f99d39057e

gherceg replied:
oh lol 0aa748286c0d4c56fedf8c7266d1af309cbbebf5

millerdev commented on corehq/apps/users/device_rate_limiter.py:
```diff
@@ -7,6 +8,9 @@
 from corehq.util.metrics import metrics_counter, metrics_histogram
 
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
```
I think this INFO level setting should go with logging configs in _settings.py_. Otherwise I think it is impossible to override the logging level in _localsettings.py_.

gherceg replied:
963cb4275a469669a08dd7ce1056e3f99d39057e

millerdev commented on corehq/apps/users/device_rate_limiter.py:
```diff
@@ -59,14 +59,9 @@ def rate_limit_device(self, domain, user_id, device_id):
-                buckets=[3, 6, 11, 21, 51],
-                tags={'domain': domain, 'user_id': user_id},
+            metrics_counter(
+                'commcare.devices_per_user.additional_device', tags={'domain': domain, 'user_id': user_id}
```
My understanding is that this will record a separate metric for each `(domain, user_id)` pair. Do you expect the number of pairs to be relatively low since users with a single device are excluded? How many unique pairs to we expect to see?

gherceg replied:
Yeah fair question. We are billed by unique custom metrics per hour. Based on some past queries, it looked like we had close to 3000 users using multiple devices in a day (over estimate). If we average that out per hour, that is 125 per hour. There is an additional variance since the `host` tag could change for the same `user_id`. So I'd maybe anticipate a bump of 200 custom metrics per hour from this metric, which is substantial, but within our budget (for now). I think in order to get an idea of actual numbers, we should start here.

gherceg committed:
Address minor feedback

gherceg committed:
Convert xform to json if instance_json IS None
